### PR TITLE
GUI for Global Illumination (VCT / CI VCT)

### PR DESCRIPTION
**Pull request by darksylinc**
# 🎉 New feature

Related to gazebosim/gz-rendering#675
Related to gazebosim/gz-rendering#435

## Summary

This PR, which depends on PR gazebosim/gz-rendering#675 to work, adds a GUI to enable and configure the new Global Illumination features added to gz-rendering

The GUI is fully functional, but there is more features planned for the future. GI is important for sensors, and right now server-side GI cannot be controlled.

This can be more complex to solve because GUI & Server should be in sync; at least in settings. And it's important the server is deterministic.

CI VCT will require attention to some specifics because the GI quality is concentric to a given camera location; hence this needs to be accounted properly when there's multiple sensors.

Realtime GI is a difficult problem to solve hence this PR introduces users to the feature. We expect to be a few bumps and a lot of feedback to fix the problems we all (users and us) encounter.

All graphics engines are in a similar situation.

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+    /// \brief See rendering::CiVctCascade::SetAreaHalfSize
+    /// \param[in] _enabled See CiVctCascade::SetAreaHalfSize
+    public: Q_INVOKABLE void SetAreaHalfSizeX(const float _x)
+        EXCLUDES(serviceMutex);
+
+    /// \brief See rendering::CiVctCascade::AreaHalfSize
+    /// \return See rendering::CiVctCascade::AreaHalfSize
+    public: Q_INVOKABLE float AreaHalfSizeX() const
+        EXCLUDES(serviceMutex);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh:160]:
```suggestion
    /// \brief See rendering::CiVctCascade::AreaHalfSizeX
    /// \return See rendering::CiVctCascade::AreaHalfSizeX
    public: Q_INVOKABLE float AreaHalfSizeX() const
        EXCLUDES(serviceMutex);
```

File: src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+    /// \brief See rendering::CiVctCascade::SetAreaHalfSize
+    /// \param[in] _enabled See CiVctCascade::SetAreaHalfSize
+    public: Q_INVOKABLE void SetAreaHalfSizeY(const float _x)
+        EXCLUDES(serviceMutex);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh:165]: Fix this one and the others below

File: src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.qml [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    minimumValue: 0.01
+    maximumValue: 100000
+    decimals: 2
+    onEditingFinished: {
+      cascadePtr.areaHalfSizeZ = value
+    }
+  }
+}
+
+/*##^##
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.qml:250]: remove?

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    /// \param[in] _ahsY Area Half Size Y
+    /// \param[in] _ahsZ Area Half Size Z
+    /// \param[in] _thinWallCounter thinWallCounter parameter
+    signals: void qmlAddCascade2(quint32 _resX, quint32 _resY, quint32 _resZ,
+                                 quint32 _octX, quint32 _octY, quint32 _octZ,
+                                 float _ahsX, float _ahsY, float _ahsZ,
+                                 float _thinWallCounter);
+
+    /// \brief See rendering::GlobalIlluminationCiVct::SetBounceCount
+    /// \param[in] _enabled See GlobalIlluminationCiVct::SetBounceCount
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh:166]:
```suggestion
    /// \param[in] _bounces See GlobalIlluminationCiVct::SetBounceCount
```

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    public: Q_INVOKABLE void SetBounceCount(const uint32_t _bounces);
+
+    /// \brief See rendering::GlobalIlluminationCiVct::BounceCount
+    /// \return See rendering::GlobalIlluminationCiVct::BounceCount
+    public: Q_INVOKABLE uint32_t BounceCount() const;
+
+    /// \brief See rendering::GlobalIlluminationCiVct::SetHighQuality
+    /// \param[in] _enabled See GlobalIlluminationCiVct::SetHighQuality
+    public: Q_INVOKABLE void SetHighQuality(const bool _quality);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh:175]:
```suggestion
    /// \param[in] _quality See GlobalIlluminationCiVct::SetHighQuality
    public: Q_INVOKABLE void SetHighQuality(const bool _quality);
```

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+    /// \brief See rendering::GlobalIlluminationCiVct::HighQuality
+    /// \return See rendering::GlobalIlluminationCiVct::HighQuality
+    public: Q_INVOKABLE bool HighQuality() const;
+
+    /// \brief See rendering::GlobalIlluminationCiVct::SetAnisotropic
+    /// \param[in] _enabled See GlobalIlluminationCiVct::SetAnisotropic
+    public: Q_INVOKABLE void SetAnisotropic(const bool _anisotropic);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh:183]:
```suggestion
    /// \brief See rendering::GlobalIlluminationCiVct::SetAnisotropic
    /// \param[in] _anisotropic See GlobalIlluminationCiVct::SetAnisotropic
    public: Q_INVOKABLE void SetAnisotropic(const bool _anisotropic);
```

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+    /// \brief See rendering::GlobalIlluminationCiVct::Anisotropic
+    /// \return See rendering::GlobalIlluminationCiVct::Anisotropic
+    public: Q_INVOKABLE bool Anisotropic() const;
+
+    /// \brief See rendering::GlobalIlluminationCiVct::SetDebugVisualizationMode
+    /// \param[in] _enabled
+    /// See GlobalIlluminationCiVct::SetDebugVisualizationMode
+    public: Q_INVOKABLE void SetDebugVisualizationMode(const uint32_t _visMode);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh:192]: Docs
```suggestion
    /// \brief See rendering::GlobalIlluminationCiVct::SetDebugVisualizationMode
    /// \param[in] _visMode
    /// See GlobalIlluminationCiVct::SetDebugVisualizationMode
    public: Q_INVOKABLE void SetDebugVisualizationMode(const uint32_t _visMode);
```

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+
+    /// \brief See rendering::GlobalIlluminationCiVct::DebugVisualizationMode
+    /// \return See rendering::GlobalIlluminationCiVct::DebugVisualizationMode
+    public: Q_INVOKABLE uint32_t DebugVisualizationMode() const;
+
+    /// \brief Binds the given camera as active for the center of all cascades
+    public: Q_INVOKABLE void OnCamareBind(const QString &_cameraName);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh:199]: doc arg

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    /// \brief Adds a new cascade based on the previous one (if there's any)
+    /// \return A CiVctCascade ptr to be used by QML
+    public: Q_INVOKABLE QObject* AddCascade();
+
+    /// \brief Pops the last created cascade
+    public: Q_INVOKABLE void PopCascade();
+
+    /// \brief Retrieves an existing cascade
+    /// \return A CiVctCascade ptr to be used by QML
+    public: Q_INVOKABLE QObject* GetCascade(int _idx) const;
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh:219]:
```suggestion
    /// \brief Retrieves an existing cascade
    /// \param[in] _idx
    /// \return A CiVctCascade ptr to be used by QML
    public: Q_INVOKABLE QObject* GetCascade(int _idx) const;
```

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.qml [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+
+  Button {
+    id: removeCascade
+    text: qsTr("Remove Cascade")
+    enabled: GlobalIlluminationCiVct.cascadesEditable
+    Layout.columnSpan: 3
+    Layout.fillWidth: true
+    onClicked: {
+      if(cascades.length > 0) {
+        //mainGridLayout.height = 400 + 400 * (cascades.length + 1)
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.qml:78]: remove?

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.qml [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+      GlobalIlluminationCiVct.OnCamareBind(textAt(currentIndex));
+    }
+    ToolTip.visible: hovered
+    ToolTip.delay: tooltipDelay
+    ToolTip.timeout: tooltipTimeout
+    ToolTip.text: qsTr("Camera around which all cascades are centered from")
+  }
+}
+
+/*##^##
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.qml:216]: remove

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.cc [hunk 12]
```diff
@@ -0,0 +1,858 @@
...
+      else
+      {
+        auto parent = baseEntity;
+        bool success = false;
+        for (size_t i = 0u; i < lidarURIVec.size() - 1; i++)
+        {
+          auto children =
+            _ecm.EntitiesByComponents(components::ParentEntity(parent));
+          bool foundChild = false;
+          for (auto child : children)
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.cc:556]:
```suggestion
          for (auto &child : children)
```

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+    public: Q_INVOKABLE void UpdateDebugVisualizationMode(int _mode);
+
+    /// \brief Set VCT resolution
+    /// \param[in] _axis Axis (width, height, depth). In range [0; 3)
+    /// \param[in] _res New resolution
+    public: Q_INVOKABLE void UpdateResolution(int _axis, uint32_t _res);
+
+    /// \brief Set VCT octant count
+    /// \param[in] _axis Axis (width, height, depth). In range [0; 3)
+    /// \param[in] _res New octant count
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh:160]:
```suggestion
    /// \param[in] _visMode New octant count
```

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+
+    /// \brief Notify fast-to-rebuild properties may have changed
+    signals: void LightingChanged();
+
+    /// \brief Notify debug visualization has changed
+    signals: void DebugVisualizationModeChanged();
+
+    /// \brief See rendering::GlobalIlluminationVct::SetResolution
+    /// \param[in] _enabled See GlobalIlluminationVct::SetResolution
+    public: Q_INVOKABLE void SetResolutionX(const uint32_t _res);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh:190]:
```suggestion
    /// \param[in] _res See GlobalIlluminationVct::SetResolution
    public: Q_INVOKABLE void SetResolutionX(const uint32_t _res);
```

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+
+    /// \brief See rendering::GlobalIlluminationVct::Resolution
+    /// \return See rendering::GlobalIlluminationVct::Resolution
+    public: Q_INVOKABLE uint32_t ResolutionX() const;
+
+    /// \brief See rendering::GlobalIlluminationVct::SetResolution
+    /// \param[in] _enabled See GlobalIlluminationVct::SetResolution
+    public: Q_INVOKABLE void SetResolutionY(const uint32_t _res);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh:198]:
```suggestion
    /// \param[in] _res See GlobalIlluminationVct::SetResolution
    public: Q_INVOKABLE void SetResolutionY(const uint32_t _res);
```

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+
+    /// \brief See rendering::GlobalIlluminationVct::Resolution
+    /// \return See rendering::GlobalIlluminationVct::Resolution
+    public: Q_INVOKABLE uint32_t ResolutionY() const;
+
+    /// \brief See rendering::GlobalIlluminationVct::SetResolution
+    /// \param[in] _enabled See GlobalIlluminationVct::SetResolution
+    public: Q_INVOKABLE void SetResolutionZ(const uint32_t _res);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh:206]:
```suggestion
    /// \param[in] _res See GlobalIlluminationVct::SetResolution
    public: Q_INVOKABLE void SetResolutionZ(const uint32_t _res);
```

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+
+    /// \brief See rendering::GlobalIlluminationVct::Resolution
+    /// \return See rendering::GlobalIlluminationVct::Resolution
+    public: Q_INVOKABLE uint32_t ResolutionZ() const;
+
+    /// \brief See rendering::GlobalIlluminationVct::SetOctantCount
+    /// \param[in] _enabled See GlobalIlluminationVct::SetOctantCount
+    public: Q_INVOKABLE void SetOctantCountX(const uint32_t _res);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh:214]:
```suggestion
    /// \param[in] _res See GlobalIlluminationVct::SetOctantCount
    public: Q_INVOKABLE void SetOctantCountX(const uint32_t _res);
```

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+
+    /// \brief See rendering::GlobalIlluminationVct::OctantCount
+    /// \return See rendering::GlobalIlluminationVct::OctantCount
+    public: Q_INVOKABLE uint32_t OctantCountX() const;
+
+    /// \brief See rendering::GlobalIlluminationVct::SetOctantCount
+    /// \param[in] _enabled See GlobalIlluminationVct::SetOctantCount
+    public: Q_INVOKABLE void SetOctantCountY(const uint32_t _res);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh:222]:
```suggestion
    /// \param[in] _res See GlobalIlluminationVct::SetOctantCount
    public: Q_INVOKABLE void SetOctantCountY(const uint32_t _res);
```

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+
+    /// \brief See rendering::GlobalIlluminationVct::OctantCount
+    /// \return See rendering::GlobalIlluminationVct::OctantCount
+    public: Q_INVOKABLE uint32_t OctantCountY() const;
+
+    /// \brief See rendering::GlobalIlluminationVct::SetOctantCount
+    /// \param[in] _enabled See GlobalIlluminationVct::SetOctantCount
+    public: Q_INVOKABLE void SetOctantCountZ(const uint32_t _res);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh:230]:
```suggestion
    /// \param[in] _res See GlobalIlluminationVct::SetOctantCount
    public: Q_INVOKABLE void SetOctantCountZ(const uint32_t _res);
```

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh [hunk 20]
```diff
@@ -0,0 +1,291 @@
...
+
+    /// \brief See rendering::GlobalIlluminationVct::OctantCount
+    /// \return See rendering::GlobalIlluminationVct::OctantCount
+    public: Q_INVOKABLE uint32_t OctantCountZ() const;
+
+    /// \brief See rendering::GlobalIlluminationVct::SetBounceCount
+    /// \param[in] _enabled See GlobalIlluminationVct::SetBounceCount
+    public: Q_INVOKABLE void SetBounceCount(const uint32_t _bounces);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh:238]:
```suggestion
    /// \param[in] _bounces See GlobalIlluminationVct::SetBounceCount
    public: Q_INVOKABLE void SetBounceCount(const uint32_t _bounces);
```

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+
+    /// \brief See rendering::GlobalIlluminationVct::BounceCount
+    /// \return See rendering::GlobalIlluminationVct::BounceCount
+    public: Q_INVOKABLE uint32_t BounceCount() const;
+
+    /// \brief See rendering::GlobalIlluminationVct::SetHighQuality
+    /// \param[in] _enabled See GlobalIlluminationVct::SetHighQuality
+    public: Q_INVOKABLE void SetHighQuality(const bool _quality);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh:246]:
```suggestion
    /// \param[in] _quality See GlobalIlluminationVct::SetHighQuality
    public: Q_INVOKABLE void SetHighQuality(const bool _quality);
```

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+
+    /// \brief See rendering::GlobalIlluminationVct::HighQuality
+    /// \return See rendering::GlobalIlluminationVct::HighQuality
+    public: Q_INVOKABLE bool HighQuality() const;
+
+    /// \brief See rendering::GlobalIlluminationVct::SetAnisotropic
+    /// \param[in] _enabled See GlobalIlluminationVct::SetAnisotropic
+    public: Q_INVOKABLE void SetAnisotropic(const bool _anisotropic);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh:254]:
```suggestion
    /// \param[in] _anisotropic See GlobalIlluminationVct::SetAnisotropic
    public: Q_INVOKABLE void SetAnisotropic(const bool _anisotropic);
```

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh [hunk 23]
```diff
@@ -0,0 +1,291 @@
...
+
+    /// \brief See rendering::GlobalIlluminationVct::Anisotropic
+    /// \return See rendering::GlobalIlluminationVct::Anisotropic
+    public: Q_INVOKABLE bool Anisotropic() const;
+
+    /// \brief See rendering::GlobalIlluminationVct::SetConserveMemory
+    /// \param[in] _enabled See GlobalIlluminationVct::SetConserveMemory
+    public: Q_INVOKABLE void SetConserveMemory(const bool _conserveMemory);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh:262]:
```suggestion
    /// \param[in] _conserveMemory See GlobalIlluminationVct::SetConserveMemory
    public: Q_INVOKABLE void SetConserveMemory(const bool _conserveMemory);
```

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh [hunk 24]
```diff
@@ -0,0 +1,291 @@
...
+
+    /// \brief See rendering::GlobalIlluminationVct::ConserveMemory
+    /// \return See rendering::GlobalIlluminationVct::ConserveMemory
+    public: Q_INVOKABLE bool ConserveMemory() const;
+
+    /// \brief See rendering::GlobalIlluminationVct::SetThinWallCounter
+    /// \param[in] _enabled See GlobalIlluminationVct::SetThinWallCounter
+    public: Q_INVOKABLE void SetThinWallCounter(const float _thinWallCounter);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh:270]:
```suggestion
    /// \param[in] _thinWallCounter See GlobalIlluminationVct::SetThinWallCounter
    public: Q_INVOKABLE void SetThinWallCounter(const float _thinWallCounter);
```

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh [hunk 25]
```diff
@@ -0,0 +1,291 @@
...
+
+    /// \brief See rendering::GlobalIlluminationVct::ThinWallCounter
+    /// \return See rendering::GlobalIlluminationVct::ThinWallCounter
+    public: Q_INVOKABLE float ThinWallCounter() const;
+
+    /// \brief See rendering::GlobalIlluminationVct::SetDebugVisualizationMode
+    /// \param[in] _enabled See GlobalIlluminationVct::SetDebugVisualizationMode
+    public: Q_INVOKABLE void SetDebugVisualizationMode(const uint32_t _visMode);
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh:278]:
```suggestion
    /// \param[in] _visMode See GlobalIlluminationVct::SetDebugVisualizationMode
    public: Q_INVOKABLE void SetDebugVisualizationMode(const uint32_t _visMode);
```

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.qml [hunk 26]
```diff
@@ -0,0 +1,295 @@
...
+import QtQuick 2.9
+import QtQuick.Controls 2.1
+import QtQuick.Controls.Material 2.2
+import QtQuick.Layouts 1.3
+//import QtQuick.Controls.Styles 1.4
+import "qrc:/qml"
+
+/*
+import "qrc:/ComponentInspector"
+import "qrc:/"*/
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.qml:27]: remove

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.qml [hunk 27]
```diff
@@ -0,0 +1,295 @@
...
+    onCurrentIndexChanged: {
+      if (currentIndex < 0|| currentIndex > 4) {
+        return;
+      }
+      GlobalIlluminationVct.debugVisualizationMode = currentIndex
+    }
+  }
+}
+
+/*##^##
```
Comment (ahcorde) [inline src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.qml:290]: remove ?

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Comment by chapulina** [comment]
Thanks for the plugin! I only looked at the code superficially, and it looks to me that it doesn't have any `gz-sim` specific functionality, is that the case? If so, this plugin could be in `gz-gui` instead, so applications other than the simulator can leverage it.

**Comment by darksylinc** [comment]
> it looks to me that it doesn't have any gz-sim specific functionality, is that the case? If so, this plugin could be in gz-gui instead

Right now it doesn't, but it is expected to in the future.

As explained in the PR description right now the plugin will only enable GI **_in the preview;_** so you can see it "pretty" on the GUI. But it is expected in the future that this plugin will control GI server-side as well.

This is a bit complex to debate because there are different issues to address in how server & gui client are synchronized:

 - Option 1: Server & Client sync their settings. Client & Server may see different things due to various differences (like camera placement for the case of CI VCT, **_when_** was the VCT internal structure build)
 - Option 2: Server & Client sync VCT's internal structures. This may be BW intensive and too hard to do; but Server & Client would see exactly the same
 - Option 3: Very little sync; only a few settings. In this case what both see may diverge a lot

Our aim is at option 1; but right now this PR only introduces GUI-only GI.

File: src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh [hunk 28]
```diff
@@ -0,0 +1,265 @@
...
+
+    Q_PROPERTY(
+      float thinWallCounter
+      READ ThinWallCounter
+      WRITE SetThinWallCounter
+      // NOTIFY LightingChanged
+    )
+
+    /// \brief Constructor
+    public: CiVctCascadePrivate(std::mutex &_serviceMutex,
```
[the remark this answers is not in the document]
Comment (darksylinc) [reply src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh:144]: Done

File: src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh [hunk 29]
```diff
@@ -0,0 +1,265 @@
...
+
+    /// \brief Set VCT octant count
+    /// \param[in] _axis Axis (width, height, depth). In range [0; 3)
+    /// \param[in] _res New octant count
+    public: Q_INVOKABLE void UpdateOctantCount(int _axis, uint32_t _count)
+        EXCLUDES(serviceMutex);
+
+    /// \brief Set the area half size
+    /// \param[in] _axis Axis (width, height, depth). In range [0; 3)
+    /// \param[in] _res New size for that axis
```
[the remark this answers is not in the document]
Comment (darksylinc) [reply src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh]: Done

File: src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh (hunk 1 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh]: Those are not typos. **rendering**::CiVctCascade::AreaHalfSize is the correct function to look for.
I've added extra doc to make this clearer, hopefully

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh (hunk 24 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh]: Done.

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh (hunk 25 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh]: Done.

File: src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh [hunk 30]
```diff
@@ -0,0 +1,265 @@
...
+
+    /// \brief Set VCT resolution
+    /// \param[in] _axis Axis (width, height, depth). In range [0; 3)
+    /// \param[in] _res New resolution
+    public: Q_INVOKABLE void UpdateResolution(int _axis, uint32_t _res)
+        EXCLUDES(serviceMutex);
+
+    /// \brief Set VCT octant count
+    /// \param[in] _axis Axis (width, height, depth). In range [0; 3)
+    /// \param[in] _res New octant count
```
[the remark this answers is not in the document]
Comment (darksylinc) [reply src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh:159]: Done.

File: src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh [hunk 31]
```diff
@@ -0,0 +1,265 @@
...
+    public: Q_INVOKABLE void UpdateAreaHalfSize(int _axis, float _halfSize)
+        EXCLUDES(serviceMutex);
+
+    /// \brief Notify various properties may have changed
+    signals: void SettingsChanged();
+
+    /// \brief See rendering::CiVctCascade::SetAreaHalfSize
+    /// \param[in] _enabled See CiVctCascade::SetAreaHalfSize
+    public: Q_INVOKABLE void SetAreaHalfSizeX(const float _x)
+        EXCLUDES(serviceMutex);
```
[the remark this answers is not in the document]
Comment (darksylinc) [reply src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh]: Done.

File: src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh (hunk 2 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.hh]: Fixed wrong `_enabled`
rendering::CiVctCascade::SetAreaHalfSize (without the letter) is not a typo but was clarified.

File: src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.qml (hunk 3 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_civct/CiVctCascadePrivate.qml:250]: This is added by QtDesigner.

I can remove it but it keeps coming back. Should we leave it?

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh (hunk 4 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh]: Done.

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh (hunk 5 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh]: Done.

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh (hunk 7 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh]: Done.

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.qml (hunk 11 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.qml:216]: Same here, QtDesigner keeps adding this back. Should we keep it?

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.cc (hunk 12 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.cc:556]: It was dead code. Thanks! Removed.

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh (hunk 13 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh:176]: Done.

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh (hunk 14 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh]: Done.

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh (hunk 16 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh]: Done.

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh (hunk 18 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh]: Done.

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh (hunk 20 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh]: Done.

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh (hunk 21 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh]: Done.

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh (hunk 19 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh]: Done.

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh (hunk 23 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh]: Done.

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.qml (hunk 26 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.qml:27]: Done.

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh (hunk 6 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh]: Done.

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh (hunk 15 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh]: Done.

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh (hunk 17 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh]: Done.

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh (hunk 22 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.hh]: Done.

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh (hunk 9 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh]: Done.

File: src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh (hunk 8 above)
Comment (darksylinc) [reply src/gui/plugins/global_illumination_civct/GlobalIlluminationCiVct.hh]: Done.

**Comment by chapulina** [comment]
The PR this depends on, https://github.com/gazebosim/gz-rendering/pull/675, didn't get into Garden. I'll remove the label. This PR can be merged into `main`, which will become `gz-sim8` / Gazebo H.

**Comment by azeey** [comment]
waiting on version bumps

**Comment by iche033** [comment]
@osrf-jenkins run tests please

File: src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.qml (hunk 27 above)
Comment (iche033) [reply src/gui/plugins/global_illumination_vct/GlobalIlluminationVct.qml]: removed.  f788bf5

**Review by iche033 — APPROVED** [review]

**Review by ahcorde — APPROVED** [review]

**Pull request merged**